app/src/main/kotlin/org/kotlinlsp/buildsystem/Gradle.kt

```diff
@@ -1,13 +1,115 @@
 package org.kotlinlsp.buildsystem
 
+import com.intellij.mock.MockProject
+import org.gradle.tooling.GradleConnector
+import org.gradle.tooling.events.OperationType
+import org.gradle.tooling.model.idea.IdeaProject
+import org.gradle.tooling.model.idea.IdeaSingleEntryLibraryDependency
+import org.jetbrains.kotlin.analysis.api.KaImplementationDetail
 import org.jetbrains.kotlin.analysis.api.projectStructure.KaModule
+import org.jetbrains.kotlin.cli.jvm.compiler.KotlinCoreApplicationEnvironment
+import org.jetbrains.kotlin.config.JvmTarget
+import org.jetbrains.kotlin.config.LanguageVersion
+import org.kotlinlsp.analysis.services.modules.LibraryModule
+import org.kotlinlsp.analysis.services.modules.SourceModule
+import org.kotlinlsp.common.debug
+import java.io.File
+import java.nio.file.Path
+import kotlin.io.path.Path
 
-class GradleBuildSystem: BuildSystem {
-    override val markerFiles: List<String>
-        get() = listOf()    // TODO
+class GradleBuildSystem(
+    private val project: MockProject,
+    private val appEnvironment: KotlinCoreApplicationEnvironment,
+    private val rootFolder: String
+) : BuildSystem {
+    override val markerFiles: List<String> = listOf("$rootFolder/build.gradle", "$rootFolder/build.gradle.kts")
 
+    @OptIn(KaImplementationDetail::class)
     override fun resolveRootModuleIfNeeded(cachedVersion: String?): Pair<KaModule, String>? {
-        // TODO
+        val connection = GradleConnector.newConnector()
+            .forProjectDirectory(File(rootFolder))
+            .connect()
+
+        val model = connection.model(IdeaProject::class.java)
+
+        model.addProgressListener({ debug("[GRADLE] ${it.displayName}") }, OperationType.PROJECT_CONFIGURATION)
+
+        val ideaProject = model.get()
+
+        val jvmTarget = checkNotNull(JvmTarget.fromString(ideaProject.jdkName)) { "Unknown jdk target" }
+        val jdkModule = getJdkHomePathFromSystemProperty()?.let { jdkHome ->
+            LibraryModule(
+                appEnvironment = appEnvironment,
+                mockProject = project,
+                roots = listOf(jdkHome),
+                javaVersion = jvmTarget,
+                isJdk = true,
+                name = "JDK ${jvmTarget.description}"
+            )
+        }
+
+        val kotlinStdlib = LibraryModule(
+            appEnvironment = appEnvironment,
+            mockProject = project,
+            name = "Kotlin stdlib",
+            javaVersion = jvmTarget,
+            roots = listOf(getKotlinJvmStdlibJarPath()),
+        )
+
+
+        val modules = ideaProject.modules.map { module ->
```

Comment:
It is something we can figure out on another PR, but right now the dependency graph is flat, we'd need to add the dependencies between libraries (for example in the case a library depends on another library but that library is not exposed to the root module). Also I think we need that to implement services like `ModuleDependentsProvider`

Reply:
Do we need to add transitive dependencies? Gradle Tooling gives us the flat list of all direct & transitive dependencies.

Reply:
Right now it is not needed, but may be needed in the future to implement the remaining services correctly, but again I think we can let it be for now